geps/gep-1619.md

```diff
@@ -163,43 +163,299 @@ Java application servers such as Tomcat and Jetty, were the first to standardize
+- Name=_value_
 - Expires=_date_
 - Max-Age=_number_
 - Domain=_domain_
 - Path=_path-value_
 - Secure
 - HttpOnly
 
-However, there are de-facto standard attributes that have been proposed as updates to [rfc6265](https://www.rfc-editor.org/rfc/rfc6265#section-5.2):
-- SameSite=_[Strict|Lax|None]_ ([Draft RFC Update](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-cookie-same-site-00))
-- Partitioned ([Draft RFC Update](https://www.ietf.org/archive/id/draft-cutler-httpbis-partitioned-cookies-01.html))
-
-**Session Cookie Strategy**
-
-Session cookie strategy is configuration specific to the creation of the cookie by the dataplane. These are not specific configuration parameters of the cookie itself, but instead the operation to take when inserting the cookie such as:
-- Rewriting an existing cookie value
-- Inserting a new cookie
-- Prefixing an existing cookie
-- Preserving an existing cookie value
-- Only adding cookies on specific HTTP methods (POST/GET)
+Other cookie attributes not defined by RFC6265, but are captured in draft RFCs and could be considered de facto
+standards due to wide acceptance are:
+- SameSite=[Strict|Lax|None]
+- Partitioned
+
+We implement core support for the `Name`, `Path`, and `TTL` attributes as the metrics in [implementations](#Implementations)
+for session persistence shows these being the most widely adopted and supported. Subsequently, we proceed with `Domain`,
+`SameSite`, `HttpOnly`, and `Secure` as extended support, considering their partial acceptance among APIs. The
+`Partitioned` attribute lacks support in any implementation API and will be omitted. Similarly, since the `Expires`
+attribute is redundant compared to the preferred `Max-Age` attribute, it will also be skipped.
+
+```go
+// SessionPersistencePolicy provides a way to define session persistence rules
+// for a service.
+//
+// Support: Core
+type SessionPersistencePolicy struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    
+    // Spec defines the desired state of SessionPersistencePolicy.
+    Spec SessionPersistencePolicySpec `json:"spec"`
+    
+    // Status defines the current state of SessionPersistencePolicy.
+    Status SessionPersistencePolicyStatus `json:"status,omitempty"`
+}
+
+// SessionPersistencePolicySpec defines the desired state of
+// SessionPersistencePolicy.
+// Note: there is no Override or Default policy configuration.
+type SessionPersistencePolicySpec struct {
+    // TargetRef identifies an API object to apply policy to.
+    // HttpRoutes, GRPCRoute, Services are the only valid API target 
+    // references.
+    TargetRef gatewayv1a2.PolicyTargetReference `json:"targetRef"`
+    
+    // HttpCookie defines and configures a cookie to achieve
+    // session persistence. 
+    HttpCookie *HttpCookie `json:"httpCookie"`
+}
+
+// HttpCookie defines a cookie to achieve session persistence.
+//
+// Support: Core
+type HttpCookie struct {
```

Comment:
considering  that a session can be initiated either by the gateway or the backend server

> For both header-based and cookie-based sessions, either the gateway or the backend server can initiate establishing the session via the appropriate header or set-cookie attributes. 


does this API define only the case when the gateway sets the cookie? 

if so, would it make sense to explicitly mention that?
if not, how will it work with for the case when the backend server sets the cookie?

Reply:
To avoid complications we should limit the discussion here to only Gateway (or mesh) issued cookies. Backend server issued cookies (honored by the clients) is the application's own logic so not really part of this proposal?

Reply:
For Java - the cookie is generated by application ( servlet library implementation ), and gateway configuration must match the cookie and value encoding associated with the implementation. PHP does a similar thing. 
I agree this kind of persistence is outside the scope of this spec - since there is nothing in the K8S API that could impact them, it's a feature of the language/framework and LB infrastructure that is not exposed to users and not configurable outside of the infra.

I don't think the spec should require who generates the cookie - it may be gateway impl, backend (sidecar or impl)  or some other infrastructure piece.

Reply:
> I don't think the spec should require who generates the cookie - it may be gateway impl, backend (sidecar or impl) or some other infrastructure piece.

That's a good point. I added a section to address `Cookie Origination` https://github.com/kubernetes-sigs/gateway-api/compare/3e631a6e46895dee15ed90443fbc9647a889d747..d292be38dec0ef163cd3edf21dfa59e214495115

Reply:
> #### Cookie Origination
> Most implementations implement session persistence by originating cookies at gateways. This spec covers the use cases related to that particular architecture. However, it's important to note that this spec does not require cookies to originate solely at a gateway. The cookie and persistent session could originate from a backend, a sidecar in a backend, or any other infrastructure component.
>
> Cookie Rewriting[¶](https://deploy-preview-2159--kubernetes-sigs-gateway-api.netlify.app/geps/gep-1619/#cookie-rewriting)
Cookie rewriting happens when the backend generates the cookie before the gateway has the opportunity to create it, and then the gateway modifies the cookie's value and/or attributes. While this was previously described as a [use case](https://deploy-preview-2159--kubernetes-sigs-gateway-api.netlify.app/geps/gep-1619/#Use-Cases-Cookies), when implementing this specification, implementations should avoid cookie rewriting with the values specified in this API. The values in this API should only be applied when the gateway is responsible for creating the cookie.

Thanks for additional clarification. 

However, I still think it is a bit unclear. Let's consider the following case:
Let's say we have a session persistent policy with some cookie. And let's assume the backend also inserts a cookie with the same name. Does this mean (from reading Cookie Origination and Cookie Rewriting sections) that  (1) the data plane should not modify any part of that cookie  if it sees that in the response from the backend and (2) the data plane should remember to route any next request with that cookie value to that backend?
Will such behavior be considered Core by future conformance tests?

Reply:
Good point, that is confusing.

Based on @costinm's recent [comment](https://github.com/kubernetes-sigs/gateway-api/pull/2159#issuecomment-1652706066), I think he's advocating for less prescription of low-level configuration in session persistence. I think we can extend that philosophy to avoid prescribing actions/non-actions in specific scenarios.

I'm going to tweak this section to be less prescriptive on what should happen if a cookie already exists, i.e. the details should be left up to the implementations.

Thoughts? Disagree?

Reply:
@gcs278 

Two thoughts here:

(1) Would we create potential portability issues? For example, if somebody changes an implementation, and this aspect of session persistence works differently in the new implementation.

(2) I wonder if different or extended configuration is needed for the case described by @costinm 
For example, we can tell the implementation to expect that a backend creates a cookie and associate that cookie value with a particular backend (how this is done is implementation specific). This way there is no ambiguity - the implementation will always expect the cookie from the backend. Since this case is not included right now in the spec, perhaps we can explicitly support it later?

some examples of how proxies support this case:
- https://docs.aws.amazon.com/elasticloadbalancing/latest/application/sticky-sessions.html#application-based-stickiness
- https://nginx.org/en/docs/http/ngx_http_upstream_module.html#sticky_learn